src/commands.rs

```diff
@@ -24,6 +30,105 @@ pub enum Command {
     Leaf(CommandLeaf),
 }
 
+#[derive(Copy, Clone)]
+pub struct CommandTaskParseError;
+impl Display for CommandTaskParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Failed to parse a command task")
+    }
+}
+impl fmt::Debug for CommandTaskParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Failed to parse a command task")
+    }
+}
+
+/**
+ * The executable part of a command.
+ */
+#[derive(Debug, Clone)]
+pub struct CommandTask {
+    pub base: String,
+    pub variables: Vec<CommandTaskVariable>,
+}
+
+impl Display for CommandTask {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.base)
+    }
+}
+
+#[derive(Debug)]
+struct CommandTaskReplaceValuesError;
+impl std::error::Error for CommandTaskReplaceValuesError {}
+impl fmt::Display for CommandTaskReplaceValuesError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Could not replace the placeholders in the CommandTask string"
+        )
+    }
+}
+
+impl FromStr for CommandTask {
+    type Err = CommandTaskParseError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let template_regex = Regex::new(r"\{\{(.+?)\}\}").unwrap();
+
+        let mut variables = vec![];
+        for template_data in template_regex.captures_iter(value) {
+            variables.push(CommandTaskVariable {
+                name: template_data[1].into(),
+                // TODO Add default value parsing
+                default_value: None,
+            });
+        }
+
+        Ok(CommandTask {
+            base: value.into(),
+            variables: variables,
+        })
+    }
+}
+
+impl CommandTask {
+    pub fn to_executable_string(
+        &self,
+        variables: HashMap<String, String>,
+    ) -> Result<String, Box<std::error::Error>> {
+        let mut output = self.base.clone();
+        for task_variable in &self.variables {
+            let value = variables
+                .get(&task_variable.name)
+                .ok_or(CommandTaskReplaceValuesError)?;
+            let match_string = format!("\\{{\\{{{}\\}}\\}}", task_variable.name);
+            let re = RegexBuilder::new(&match_string).build()?;
+            output = re.replace_all(&output, value.as_str()).to_string();
```

Comment:
Well, in that case how would I return the error from the closure?
E.g. 

```rust
    pub fn to_executable_string(<···>) -> Result<String, Box<std::error::Error>> {
        let template_regex = Regex::new(r"\{\{(.+?)\}\}").unwrap();
        output = template_regex
            .replace_all(&output, |caps: &Captures| {
                variables.get(&caps[0]).ok_or(CommandTaskReplaceValuesError)? // <- Error
            })
            .to_string();
    // <···>
```

Does not compile because the error now gets returned inside the closure/.

Reply:
Yeah, you can't... you could first check that all the variables have values, and then do the actual replacing, though

Reply:
is it actually possible that variables are missing from the hash map though?